design/keycloak.x/storage-persistence.md

```diff
@@ -0,0 +1,530 @@
+   of version _N-1_ in the cluster are gradually replaced
+   by newer version. It also provides means to revert to
+   the previous version if anything goes wrong.
+
+3. _Extended forward compatibility_ (optional).
+   Object stored by store of version _N_ MAY be readable by much older store,
+   i.e. store of version _M_ where _M ≤ N-2_, but the store MUST
+   throw an `IllegalArgumentException` if the object cannot
+   be reliably reconstructed from the stored data.
+
+The following matrix illustrates the above conditions. It shows whether
+an object stored in version _X_ in a store of version _Y_ is readable
+(green) or unreadable (red):
+
+![Matrix shows whether a stored object in format _X_ is readable (green) or unreadable (red) by store of version _Y_](img/storage-persistence-read-compatibility.svg)
+
+For the sake of simplicity, object versions are treated as a sequence of integers
+thorough this text. See [below](#_otherVersioning) for other options.
+
+Versions of individual stores in the same server version are unrelated.
+For example, in Keycloak of version 1.2.3, versions can be 53 for user objects,
+41 for clients, and 14 for realms. Versioning of Keycloak and the logical objects
+is independent, so it is possible that the Keycloak version 1.2.4 may have
+the same object versions as in 1.2.3, that one of them has changed,
+or all have changed.
+
+*Note on testability:* It must be possible to automatically test that
+a store implementation satisfies all the requirements stated in this section.
+For this purpose, there is
+[a project in keycloak-playground](https://github.com/keycloak/keycloak-playground/no-downtime-upgrade/) repository
+that captures all the requirements and provides a harness to implement
+a proposal for a 0-downtime storage implementation.
+
+### <a name="_cir"></a> Custom implementation requirements (CIR)
+
+The requirements coming from the need to easily provide
+custom implementations are vague by nature but useful to state them here.
+
+1. Store interfaces must be simple enough to provide fast custom implementations.
+2. Store interfaces must be generic enough to fit any storage technology.
+3. Store interfaces must operate on logical level.
+
+## Design
+
+Actual storage can be based on various technologies. For example,
+the storage can use a file system, relational database, LDAP,
+in-memory streaming database, or a REST service.
+For this reason, it is important to keep the storage interface
+on the level of logical objects (i.e. models)
+and detached from the physical implementation.
+
+It is obvious that no-downtime requirement and performance are mutually
+outbalancing each other. Hence it should be stated that:
+
+* No-downtime constraints must be satisfied
+* Performance needs to be good-enough. This means that:
+  * Object lookup by an object identifier must succeed
+    in logarithmic time at worst
+  * Lookup by searchable attribute should be bound
+    by logarithmic time
```

Comment:
I suppose the "logarithmic time" is meant to be compared to the total count of existing objects in the store, right? Probably just a nitpick, but it may not be 100% obvious from the text IMO :-)

Reply:
Yup, will update the design